lib/rust/flexer/tests/flexer_lexer_definition_test.rs

```diff
@@ -0,0 +1,229 @@
+    }
+
+    // TODO [AA] Parent groups.
+    pub fn define_group(&mut self,name:&str) -> &mut Group {
+        let id    = self.groups.len();
+        let group = Group::new(id,String::from(name),None);
+        self.groups.push(group);
+        self.groups.get_mut(id).expect("Has just been pushed so should always exist.")
+    }
+
+    pub fn specialize(&self) -> String {
+        let group_nfa:Vec<NFA>  = self.groups.iter().map(|group|group.into()).collect();
+        let _group_dfa:Vec<DFA> = group_nfa.iter().map(|nfa|nfa.into()).collect();
+        let str = String::new();
+        str
+    }
+}
+
+#[derive(Debug)]
+#[allow(missing_docs)]
+pub struct TestLexer<Reader:LazyReader> {
+    lexer: Flexer<TestState,AST,Reader>
+}
+
+/// The definition?
+#[allow(missing_docs)]
+impl<Reader:LazyReader> TestLexer<Reader> {
+    pub fn new(reader:Reader) -> Self {
+        let lexer = Flexer::new(reader);
+        TestLexer{lexer}
+    }
+}
+
+/// Implementation helpers.
+#[allow(missing_docs)]
+impl<Reader:LazyReader> TestLexer<Reader> {
+    pub fn def_on_first_word_str(&mut self) {
+        let str = self.current_match.clone();
+        let ast = AST::Word(str);
+        self.def_on_first_word(ast);
+    }
+
+    pub fn def_on_first_word(&mut self,ast:AST) {
+        self.tokens.push(ast);
+        let id = self.seen_first_word_group.id;
+        self.begin_state(id);
+    }
+
+    pub fn def_on_spaced_word_str(&mut self) {
+        let str = self.current_match.clone();
+        let ast = AST::Word(String::from(str.trim()));
+        self.def_on_spaced_word(ast);
+    }
+
+    pub fn def_on_spaced_word(&mut self,ast:AST) {
+        self.tokens.push(ast);
+    }
+
+    pub fn def_on_err_suffix_first_word(&mut self) {
+        let ast = AST::Unrecognised(self.current_match.clone());
+        self.tokens.push(ast);
+    }
+
+    pub fn def_on_err_suffix(&mut self) {
+        self.def_on_err_suffix_first_word();
+        self.end_state();
+    }
+
+    pub fn def_on_no_err_suffix_first_word(&mut self) {}
+
+    pub fn def_on_no_err_suffix(&mut self) {
+        self.def_on_no_err_suffix_first_word();
+        self.end_state();
+    }
+}
+
+// === Trait Impls ===
+
+impl<Reader:LazyReader> Deref for TestLexer<Reader> {
+    type Target = Flexer<TestState,AST,Reader>;
+    fn deref(&self) -> &Self::Target {
+        &self.lexer
+    }
+}
+
+impl<Reader:LazyReader> DerefMut for TestLexer<Reader> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.lexer
+    }
+}
+
+// TODO [AA] _Where_ do I want to write the definition?
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub struct TestState {
+    root_group: Rc<Group>,
+    seen_first_word_group: Rc<Group>,
+    matched_bookmark: BookmarkId
+}
+
+impl<Reader:LazyReader> FlexerState<Reader> for TestState {
+    fn new(reader: &mut Reader) -> Self {
+        // TODO [AA] Parent groups need to be mutable as held.
+        let root_group = Rc::new(Group::new(0,String::from("ROOT"),None));
```

Comment:
Gropup constructor sohuld have id auto-generated. User should ofc not provide it.